torchmultimodal/models/mdetr.py

```diff
@@ -0,0 +1,676 @@
+            value=memory,
+            attn_mask=memory_mask,
+            key_padding_mask=memory_key_padding_mask,
+        )[0]
+        tgt = tgt + self.dropout3(tgt2)
+        tgt = self.norm3(tgt)
+
+        # FFN
+        tgt2 = self.linear2(self.dropout(self.activation(self.linear1(tgt))))
+        tgt = tgt + self.dropout4(tgt2)
+        tgt = self.norm4(tgt)
+        return tgt
+
+
+class FeatureResizer(nn.Module):
+    """
+    This class takes as input a set of embeddings of dimension C1 and outputs a set of
+    embedding of dimension C2, after a linear transformation, dropout and normalization (LN).
+
+    Args:   input_feat_size (int): Dimension of input features.
+            output_feat_size (int): Dimension of output features.
+            dropout (float): Dropout probability for final features. Default: 0.1
+            do_ln (bool): Whether to perform layer normalization after the linear layer.
+
+    Inputs: encoder_features (Tensor): Features to be resized.
+    """
+
+    def __init__(
+        self,
+        input_feat_size: int,
+        output_feat_size: int,
+        dropout: float = 0.1,
+        do_ln: bool = True,
+    ):
+        super().__init__()
+        self.do_ln = do_ln
+        # Object feature encoding
+        self.fc = nn.Linear(input_feat_size, output_feat_size, bias=True)
+        self.layer_norm = nn.LayerNorm(output_feat_size, eps=1e-12)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, encoder_features: Tensor):
+        x = self.fc(encoder_features)
+        if self.do_ln:
+            x = self.layer_norm(x)
+        output = self.dropout(x)
+        return output
+
+
+def get_clones(module: nn.Module, n: int) -> nn.ModuleList:
+    return nn.ModuleList([deepcopy(module) for i in range(n)])
+
+
+def mdetr_transformer(
+    d_model: int = 256,
+    nhead: int = 8,
+    num_encoder_layers: int = 6,
+    num_decoder_layers: int = 6,
+    dim_feedforward: int = 2048,
+    dropout: float = 0.1,
+    return_intermediate_dec: bool = True,
+) -> MDETRTransformer:
+    return MDETRTransformer(
+        d_model=d_model,
+        nhead=nhead,
+        num_encoder_layers=num_encoder_layers,
+        num_decoder_layers=num_decoder_layers,
+        dim_feedforward=dim_feedforward,
+        dropout=dropout,
+        return_intermediate_dec=return_intermediate_dec,
+    )
+
+
+def mdetr_resnet101(
+    num_queries: int = 100,
+    num_classes: int = 255,
+    embedding_dim: int = 768,
+    transformer_d_model: int = 256,
+    transformer_nhead: int = 8,
+    transformer_encoder_layers: int = 6,
+    transformer_decoder_layers: int = 6,
+    transformer_dim_feedforward: int = 2048,
+    transformer_dropout: float = 0.1,
+    return_intermediate_dec: bool = True,
+) -> MDETR:
+    image_backbone = resnet101()
+    image_backbone = mdetr_resnet101_backbone()
+    pos_embed = PositionEmbedding2D(128, scale=2 * math.pi)
+    image_backbone.num_channels = 2048
```

Comment:
where is 2048 coming from?

Reply:
This is just the output dim of the final ResNet layer before pooling. Ref: [1](https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py#L757), [2](https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py#L206), and [3](https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py#L115)